src/V3Fork.cpp

```diff
@@ -34,12 +46,351 @@
 #include "V3Ast.h"
 #include "V3AstNodeExpr.h"
 #include "V3Global.h"
+#include "V3MemberMap.h"
 
 #include <algorithm>
+#include <map>
 #include <set>
+#include <vector>
 
 VL_DEFINE_DEBUG_FUNCTIONS;
 
+class DynScopeInstance final {
+public:
+    AstClass* classp = nullptr;  // Class for holding variables of dynamic scope
+    AstClassRefDType* refDTypep = nullptr;  // RefDType for the above
+    AstVar* handlep = nullptr;  // Class handle for holding variables of dynamic scope
+
+    // True if the instance exists
+    operator bool() const { return classp != nullptr; }
+};
+
+class DynScopeFrame final {
+public:
+    DynScopeFrame(AstNodeModule* modp, AstNode* procp)
+        : m_modp(modp)
+        , m_procp(procp) {}
+
+    DynScopeInstance& createInstancePrototype() {
+        UASSERT_OBJ(!m_instance, m_procp, "Dynamic scope already instantiated.");
+
+        m_instance.classp = new AstClass{m_procp->fileline(), generateDynScopeClassName(m_procp)};
+        m_instance.refDTypep
+            = new AstClassRefDType{m_procp->fileline(), m_instance.classp, nullptr};
+        v3Global.rootp()->typeTablep()->addTypesp(m_instance.refDTypep);
+        m_instance.handlep = new AstVar{m_procp->fileline(), VVarType::BLOCKTEMP,
+                                        generateDynScopeHandleName(m_procp), m_instance.refDTypep};
+        m_instance.handlep->funcLocal(true);
+        m_instance.handlep->lifetime(VLifetime::AUTOMATIC);
+
+        return m_instance;
+    }
+
+    const DynScopeInstance& instance() const { return m_instance; }
+    void captureVar(AstVar* varp) { m_captures.insert(varp); }
+    bool captured(AstVar* varp) { return m_captures.count(varp) != 0; }
+    AstNode* procp() const { return m_procp; }
+
+    void populateClass() {
+        UASSERT_OBJ(m_instance, m_procp, "No DynScope prototype");
+
+        // Move variables into the class
+        for (AstVar* varp : m_captures) {
+            if (VL_UNLIKELY(varp->direction() == VDirection::INPUT)) {
```

Comment:
```suggestion
            if (varp->direction() == VDirection::INPUT) {
```

I don't see why this would be unlikely.

Reply:
I just found referencing a function argument inside of a fork it to be a less common scenario than referencing anything else, especially given that to my understanding we can reference only the variables from the block that contains the fork - it can be a function but it can be a  `begin..end` block as well.